flux: add compatibility for Flux in the ATS codebase

wihobbs opened the pull request:
Per our discussion today with @dawson6, @davidbloss, and @jameshcorbett, this PR is opened to test the compatibility of a work-in-progress Flux executor and monitor the progress of expanding ATS compatibility for Flux.

Currently, the Flux executor has option compatibility for:
- number of cores
- number of nodes
- exclusivity within an allocation
- verbose output of Flux key value store

To test this on Ruby (or another slurm-runnning LC cluster), install ATS in a virtual environment and request an allocation from slurm:
```salloc -N 3 -p pdebug -A cbronze -t 60```

And then start a flux instance using `srun`:
```srun -N 3 --pty flux start```

wihobbs commented:
I'm setting the base branch as `python3` to isolate the changes that I made. When I have the okay from @dawson6 I'll work out the kinks between my branch and `main`.

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,101 @@
+        
+        # set number of nodes and cores
+        ret.append(f"-N {nn}")
+        ret.append(f"-n {np}")
```
This works with Flux's command-line parsing, but you're combining the arguments in a way that some parsers might not like. I think this should either be `ret.append(f"-n{np}")` or `ret += ["-n", str(np)]`. When you see `-N 5` on the command line, that's really two arguments---that's all I'm getting at.

davidbloss replied:
Another option: `ret.extend(["-n", str(np)])`

wihobbs replied:
Thinking about this now that I'm adding thread support. In `flux mini run` this is specified as `--threads=1` for 1 thread, etc. With the `=` is this one argument, or two? How many is it supposed to be?

jameshcorbett replied:
I have never seen that `--threads` option, where are you getting that from? Threads aren't really a resource, so AFAIK Flux doesn't allocate them. User's applications are free to create however many threads they want.

jameshcorbett replied:
When I said that `-N 5` is really two arguments, what I meant was that if you execute `./foo -N 5` in your shell, `argc` for that process will be 3. `argv[1]` would be `-N` and `argv[2]` would be `-5`.

If you want to play around with this you could try:

```bash
echo "import sys; print(sys.argv)" > script.py
bash-5.1$ python3 script.py --threads=1
['script.py', '--threads=1']
```

so as you can see `--threads=1` is one argument

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,101 @@
+    def periodicReport(self):
+        ## Copied from SlurmProcessorScheduled lines 650-654
+        "Report on current status of tasks"
+        if len(self.running):
```
If you're just checking whether `self.running` is not empty, the pythonic way to do that is just `if self.running:`. On the other hand `if len(self.running) > 0` is a little clearer... (I am ambivalent about this particular aspect of the official Python guidelines)

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,101 @@
+        ## Flux specific accessors for number of nodes
+        
+        procs = flux.resource.list.resource_list(f).get().allocated.nnodes
+        total = flux.resource.list.resource_list(f).get().free.nnodes
```
Would be better to call `resource_list` once and save the result.

wihobbs replied:
Good idea, and it would probably be better to do this in `init` as well.

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,101 @@
+        ## anything to initialize here?
+        ## set ceiling on number of nodes and cores in the allocation.
+        self.maxNodes = int(flux.resource.list.resource_list(f).get().free.nnodes)
+        self.maxCores = int(flux.resource.list.resource_list(f).get().free.ncores)
```
Same as below, would be better to call resource_list once and save the result.

wihobbs replied:
Ah :)

jameshcorbett reviewed (commented):
@wihobbs asked me to make a nitpicky review for educational purposes so here it is.

davidbloss commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,101 @@
+#ATS:slurm56 SELF FluxScheduled 3
+#ATS:toss_4_x86_64_ib SELF FluxScheduled 16
+
+from ats.atsMachines import lcMachines
```
💨 🧹  [isort](https://pycqa.github.io/isort/) is an import sorting tool that does this for you - also available as a [vscode extension](https://github.com/microsoft/vscode-isort) until you transition to [neovim](https://github.com/neovim/neovim) :wink:

wihobbs replied:
Are flux imports considered third party imports (should they go before ats imports or with ats imports?) Not sure why, but isort is trying to group them like this, which looks silly to me:
```python3
import time
from math import ceil

import flux.job
from ats import terminal
from ats.atsMachines import lcMachines

import flux
```
It turns out, command line isort and vscode isort disagree on where flux should go. CLI says this which looks nicer to me, but still not sure if there should be a line between flux and ats imports:
```python3
import time
from math import ceil

import flux
import flux.job
from ats import terminal
from ats.atsMachines import lcMachines
```

jameshcorbett replied:
Flux would definitely be a third party import in this case. Flux is a separate project from ATS. If ATS was in the flux-framework GitHub org and written by Flux developers it would be different.

So with that in mind I would say it should be 

```python
import time
from math import ceil

import flux
import flux.job

from ats import terminal
from ats.atsMachines import lcMachines
```

i.e. the second one, but with an extra line like you said.

wihobbs commented:
Thanks, @jameshcorbett, and sorry for the delay in pushing this -- still finding my way around in git. I think I addressed all of your comments (hopefully more to come ;))

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,115 @@
+## but there's no assurances it will be to the _same_ compute nodes for each of ats's "tests" or whatever
+## answer: YES for now ## Should start its own Flux session!! TODO
+
+FLUX_HANDLE = flux.Flux()  ## where do I go?
```
I would say it should be an instance attribute. It's fine as a module constant like it is, but I don't really see why it should be, and I generally think that if there isn't a reason for it to be a module-level constant or a class attribute, it should be an instance attribute.

I would say in general:

- Module constant/attribute if it needs to be shared among functions or classes
- Class attribute if it needs to be shared among instances
- Otherwise, it should be an instance attribute

Instance attributes generally go in the `__init__` method, e.g.

```python

def __init__(self, *args, **kwargs):
    self._flux_handle = flux.Flux()
    super().__init__(*args, **kwargs)
```

Although it looks like this class has an `init` method which also defines instance attributes. That's not standard but there may be good reasons for it.

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,115 @@
+        Defines a persistent handle to use to connect to the broker."""
+        self.maxNodes = int(flux.resource.list.resource_list(FLUX_HANDLE).get().up.nnodes)
+        self.maxCores = int(flux.resource.list.resource_list(FLUX_HANDLE).get().up.ncores)
+        self.coresPerNode = int(self.maxCores // self.maxNodes)
```
`int()` call is redundant here, the `//` gives you an integer.

wihobbs replied:
I generally kind of like being redundant with types in Python since Python is so loose about types. But it's removed

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,115 @@
+        """Optparse (soon argparse) parameters from command-line options
+        for ATS users. Needed for functionality with .ats files.
+        
+        :param options: The options available to a user in test.ats files."""
```
I'm not sure this is ever explicitly stated in https://peps.python.org/pep-0257/#multi-line-docstrings but I think if a docstring is multi-line the closing triple quote goes on its own line, e.g. in this case 

```
:param options: The options available to a user in test.ats files.
"""
```

Similarly for other methods.

wihobbs replied:
Seems cleaner, will fix.

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,115 @@
+        ret = "flux mini run".split()
+        np = test.options.get('np', 1)
+        nn = test.options.get('nn', 1)
+        max_time = test.options.get('timelimit', '29m')
```
Why 29 minutes?

wihobbs replied:
Default I copied over from somewhere else in the ATS codebase (tests.py I think).

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,115 @@
+            nn = ceil(np / self.coresPerNode)
+        ret.append(f"-N{nn}")
+        ret.append(f"-n{np}")
+        ret.append(f"-t{max_time}")
```
Have you checked that ATS's time limit notation is compatible with Flux's? Otherwise you might need to do some conversion (sounds kind of ugly and not fun)

jameshcorbett commented on ats/atsMachines/fluxScheduled.py:
```diff
@@ -0,0 +1,115 @@
+        self.exclusive = options.exclusive
+        self.timelimit = options.timelimit
+
+    def calculateCommandList(self, test: AtsTest):
```
This is the one place you have type annotations. I think that's fine? I guess unless ATS used mypy or another static type checker though, the annotations are just another form of documentation (and like regular documentation, can be wrong).

jameshcorbett commented on test/unit/flux/t0001-flux.py:
```diff
@@ -0,0 +1,20 @@
+
+class TestFluxScheduled(unittest.TestCase):
+    @classmethod
+    def setUp():
```
If it's a classmethod it needs to take ``cls`` as the first argument. But ``setUp`` is an instance method and so should take ``self`` as the first argument, see the example here https://docs.python.org/3/library/unittest.html#unittest.IsolatedAsyncioTestCase.run

But also I might not use ``setUp`` at all and might create an ``AtsTest()`` instance in each test function. That's up to you though, depends on how much duplication it would create.

wihobbs replied:
I think I'm leaning towards creating an `AtsTest` in each function. That way, I could test both valid and invalid ones.

jameshcorbett commented on test/unit/flux/t0001-flux.py:
```diff
@@ -0,0 +1,20 @@
+    def setUp():
+        self.test = AtsTest()
+
+    def test_command_list(self, test: AtsTest):
```
Tests don't take any arguments other than ``self``.

wihobbs replied:
Sorry, I started at the bottom and read up. Again, thinking method is the way to go here.

jameshcorbett commented on test/unit/flux/t0001-flux.py:
```diff
@@ -0,0 +1,20 @@
+        return 0
+
+    def test_create_flux_instance(self):
+        return 0
```
Tests don't need to return anything. If a test doesn't fail an assertion or raise an exception it is considered to have succeeded.

jameshcorbett reviewed (commented):
Left some comments below.

I understand why you're not sure what to test. There isn't a whole lot to test since the methods either print something or construct a list of strings. Like you suggested you could check that the list of strings is constructed correctly. I think a good test though would be to check whether the command is accepted by Flux. The way I would do that is to add the ``--dry`` option to the command, which just makes the command spit out the jobspec it would submit rather than actually submitting it. Without the flag Flux will complain if a Flux instance isn't running. If you wanted to, you could then read the jobspec that Flux constructed, and verify that the numbers check out.

You could skip that test if Flux wasn't installed, too (check ``shutil.which("flux") is not None``), otherwise it would give false failures on systems without Flux.

jameshcorbett commented:
> If you wanted to, you could then read the jobspec that Flux constructed, and verify that the numbers check out.

Upon further thought I think that's a good idea but I would actually use Flux's jobspec class to read in the jobspec that was constructed. And I had forgotten that you won't be able to import the ``FluxScheduled`` class unless you have the ``flux`` package, so the check of whether Flux is installed is somewhat redundant.

wihobbs commented:
Well groan. I'm sure I did something wrong to make the working tree this out of whack, but this _should_ be aligned with the `main` branch. 

Making a note to myself that this is what I did and in the future I should not do this again (it seems like `rebase` should do this better but I must be using it wrong)

```shell
(fluke_ats) [hobbs17@corona173:ATS]$ git config pull.rebase true
(fluke_ats) [hobbs17@corona173:ATS]$ git pull -X theirs upstream main
```

Tomorrow I'll go through `machines.py` and the commit history and make sure none of the commits from @dawson6 and @davidbloss change any of my implementation for Flux (but at first glance they shouldn't). And find a way to make my branch's commit history...less yucky.

wihobbs commented:
All right, I believe I fixed the commit so that it will not have merge conflicts and can be merged into main once WIP is dropped. Making a note here that these are the pending actions for dropping WIP:
- test GPU functionality again and make sure GPU allocation logic works for more complex cases
- finish `atsflux` binding with additional functionality

wihobbs commented:
I believe at this point I have addressed all of the must-haves for ATS/Flux but will run more tests on Monday to be sure.

wihobbs commented on ats/atsMachines/slurmProcessorScheduled.py:
```diff
@@ -6,7 +6,6 @@
 #ATS:slurm24                 SELF SlurmProcessorScheduled 24
 #ATS:slurm32                 SELF SlurmProcessorScheduled 32
 #ATS:slurm36                 SELF SlurmProcessorScheduled 36
-#ATS:slurm56                 SELF SlurmProcessorScheduled 56
```
Oh whoops. Thought I added that back in. will fix

wihobbs commented on ats/bin/atsflux.py:
```diff
@@ -0,0 +1,133 @@
+import argparse
+import os
+import sys
+import subprocess as sp
```
Sounds good to me.

jameshcorbett commented on ats/bin/atsflux.py:
```diff
@@ -0,0 +1,133 @@
+            f"-N{args.numNodes}",
+            f"-n{args.numNodes}",
+            "--pty",
+            "/usr/bin/flux",
```
I would replace this with [shutil.which("flux")](https://docs.python.org/3.8/library/shutil.html#shutil.which). By this point you're guaranteed to have Flux in PATH because you ran `flux -V` already, but there's no assurance that it will be in `/usr/bin`.

wihobbs commented:
I removed work-in-progress because I believe we are good for final review here. Thanks @dawson6, @jameshcorbett, and @davidbloss -- I believe I have addressed all of your comments from here and our meetings.

davidbloss reviewed (approved):
Looks good!
